pkg/certs/certs.go

```diff
@@ -0,0 +1,120 @@
+/*
+ * Minio Cloud Storage, (C) 2018 Minio, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package certs
+
+import (
+	"crypto/tls"
+	"sync"
+
+	"github.com/rjeczalik/notify"
+)
+
+// A Certs represents a certificate manager able to watch certificate
+// and key pairs for changes.
+type Certs struct {
+	sync.RWMutex
+	// user input params.
+	certFile string
+	keyFile  string
+	loadCert LoadX509KeyPairFunc
+
+	// points to the latest certificate.
+	cert tls.Certificate
+
+	// internal param to track for events, also
+	// used to close the watcher.
+	e chan notify.EventInfo
+}
+
+// LoadX509KeyPairFunc - provides a type for custom cert loader function.
+type LoadX509KeyPairFunc func(certFile, keyFile string) (tls.Certificate, error)
+
+// New initializes a new certs monitor.
+func New(certFile, keyFile string, loadCert LoadX509KeyPairFunc) (*Certs, error) {
+	c := &Certs{
+		certFile: certFile,
+		keyFile:  keyFile,
+		loadCert: loadCert,
+		// Make the channel buffered to ensure no event is dropped. Notify will drop
+		// an event if the receiver is not able to keep up the sending pace.
+		e: make(chan notify.EventInfo, 1),
+	}
+
+	if err := c.watch(); err != nil {
+		return nil, err
+	}
+
+	return c, nil
+}
+
+// watch starts watching for changes to the certificate
+// and key files. On any change the certificate and key
+// are reloaded. If there is an issue the loading will fail
+// and the old (if any) certificates and keys will continue
+// to be used.
+func (c *Certs) watch() (err error) {
+	if err = notify.Watch(c.certFile, c.e, eventWrite...); err != nil {
+		return err
+	}
+
+	if err = notify.Watch(c.keyFile, c.e, eventWrite...); err != nil {
+		return err
+	}
+	c.Lock()
+	c.cert, err = c.loadCert(c.certFile, c.keyFile)
```

Comment:
In addition to loading the certificate before watching, this function can automatically call `notify.Stop()` (in case one of the Watch calls succeeds before an error is encountered). This would make the certs package a bit safer/easier to use (e.g. user of this module need not call Stop in case of errors).

Reply:
if watch returns an error there is no need to notify.Stop() since the returned value of `certs` is nil anyway. The application is dead and all watches for that application have been relinquished automatically by the kernel.

Reply:
Well watch will also fail if the files don't exist @vadmeste so the end result is same.

Reply:
even if the files are corrupted this function simply returns an error during New() so subsequent caller never has to do notify.Stop()

Reply:
Well it is fine and works for now. However, we need something like:

```
defer func() {
    if (err != nil) {
        notify.Stop(c.e)
    }
}
```

at the beginning of the function to make it properly correct. Though the certs package is being used like this today, a future change may not exit the application in case of error - this might lead to a leaking watch. The certs package should not require/assume such behaviour on the application IMO.